File: ppds/savages.c
#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>

typedef struct {
  sem_t mutex, empty, full;
  int servings;
} pot;

void pot_init(pot *p) {
  sem_init(&p->mutex, 0, 1);
  sem_init(&p->empty, 0, 0);
  sem_init(&p->full, 0, 0);
  p->servings = 0;
}

void cook(pot *p) {
  for (;;) {
    sem_wait(&p->empty);
    p->servings = 10;
    printf("cooked %d servings\n", p->servings);
    sem_post(&p->full);
  }
}

void *savage(void *v) {
  pot *p = v;
  for (;;) {
    sem_wait(&p->mutex);
    if (p->servings == 0) {
      sem_post(&p->empty);
      sem_wait(&p->full);
    }
    p->servings--;
    sem_post(&p->mutex);
    
    printf("eating\n");
    sleep(1);
  }
  return NULL;
}

int main() {

  int i, n = 7;
  pot p;
  pot_init(&p);
  pthread_t th[n];

  for (i=0; i<n; i++) {
    pthread_create(&th[i], NULL, savage, &p);
  }

  cook(&p);
  return 0;
}

File: ppds/savages.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

type pot struct {
	mu          sync.Mutex
	servings    int
	empty, full chan struct{}
}

func savage(n int, p *pot) {
	for {
		// get serving from pot
		p.mu.Lock()
		if p.servings == 0 {
			p.empty <- struct{}{}
			<-p.full
		}
		p.servings--
		p.mu.Unlock()

		// eat
		fmt.Printf("%d: eating nom, nom ..\n", n)
		time.Sleep(2 * time.Second)
	}
}

func cook(p *pot) {
	for {
		// put serving in pot
		<-p.empty
		p.servings = 1 + rand.Int()%10
		fmt.Printf("puting %d in pot\n", p.servings)
		p.full <- struct{}{}
	}
}

func main() {
	var (
		n = 10
		p = &pot{
			empty: make(chan struct{}, 1),
			full: make(chan struct{}, 1),
		}
	)
	for i := 0; i < n; i++ {
		go savage(i, p)
	}

	cook(p)
}
